scale-decode/src/utils/stack_vec.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright (C) 2023 Parity Technologies (UK) Ltd. (admin@parity.io)
+// This file is a part of the scale-decode crate.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//         http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// A quick hopefully-stack-only vec implementation for holding TypeIds.
+pub struct StackVec<T, const N: usize> {
```

Comment:
would interesting with some benchmarks on this compared to `Vec` and `SmallVec` :)

Reply:
Hehe yeah; I just wrote the minimum bit really to avoid another dependency; the hope is that it will never need to allocate unless something crazy comes up!

Reply:
If we'd want to squeeze every bit of performance out of it, the current implementation could allocate twice when moving the vector from the stack to the heap.

```rust
// First allocation goes here
                    let mut v = items[..].to_vec();
                    
// Potential allocation for extending the vector + a memcpy
                    v.push(item);
                    
                    self.inner = StackVecInner::Heap { items: v };
```

We could optimize those similar to:

```rust
            let mut v = Vec::with_capacity(N + 1);
            // SAFETY: we allocate the appropriate number of elements
            unsafe {
                items.as_ptr().copy_to_nonoverlapping(v.as_mut_ptr(), N);
                v.set_len(N);
                // We could probably do some pointer magic to call `set_len` only once
                v.push(item);
            }
            self.inner = StackVecInner::Heap { items: v };
```

Please feel free to ignore this :D Seems that may not be the most common case

Reply:
I'm hoping it'll never happen; only if a compact encoded type is nested in a crazy number of new type wrappers :) so yeah it's no biggie, but nice one!